[lldb] Fix race condition in Process::WaitForProcessToStop()

athierry-oct opened the pull request:
This PR addresses a race condition encountered when using LLDB through the Python scripting interface.

I'm relatively new to LLDB, so feedback is very welcome, especially if there's a more appropriate way to address this issue.

### Bug Description

When running a script that repeatedly calls `debugger.GetListener().WaitForEvent()` in a loop, and at some point invokes `process.Kill()` from within that loop to terminate the session, a race condition can occur if `process.Kill()` is called around the same time a breakpoint is hit.

### Race Condition Details

The issue arises when the following sequence of events happens:

1. The process's **private state** transitions to `stopped` when the breakpoint is hit.
2. `process.Kill()` calls `Process::Destroy()`, which invokes `Process::WaitForProcessToStop()`. At this point:
   - `private_state = stopped`
   - `public_state = running` (the public state has not yet been updated)
3. The **public stop event** is broadcast **before** the hijack listener is installed.
4. As a result, the stop event is delivered to the **non-hijack listener**.
5. The interrupt request sent by `Process::StopForDestroyOrDetach()` is ignored because the process is already stopped (`private_state = stopped`).
6. No public stop event reaches the hijack listener.
7. `Process::WaitForProcessToStop()` hangs waiting for a public stop event, but the event is never received.
8. `process.Kill()` times out after 20 seconds

### Fix Summary

This patch modifies `Process::WaitForProcessToStop()` to ensure that any pending events in the non-hijack listener queue are processed before checking the hijack listener. This guarantees that any missed public state change events are handled, preventing the hang.


### Additional Context

A discussion of this issue, including a script to reproduce the bug, can be found here: 
 [LLDB hangs when killing process at the same time a breakpoint is hit](https://discourse.llvm.org/t/lldb-hangs-when-killing-process-at-the-same-time-a-breakpoint-is-hit)

labath commented:
Jim is the process state king here, but FWIW, I think you are onto something here -- I just think it needs to be done more carefully. Fetching the events from the non-hijack listener in this way is racy as there could be another thread trying to do the same (that's sort of why we bother with hijacking in the first place). Instead, what I think could work is to -- in the process of installing the hijack listener -- take (hijack) the *pending* events from the real listener and enqueue them to the hijack listener instead. This should be done while holding the event mutex of the real listener to make sure the other thread doesn't trample over this.

Then you should be able to listen on the hijack listener to your heart's content.

(I don't know if this would matter in practice, but for completeness, we probably should also re-enqueue any unlistened-for events from the hijack listener back to the real listener.)

jimingham commented:
Your solution is going in the right direction, but is trying to solve a problem "after the fact" that we can solve better by preventing the public stop events from making their way to the non-primary listener before the primary listener gets a chance to fully process them.

Process event listeners are a bit special because we have to keep the "process event state" in sync with the value returned from Process::GetState.  It would be hard to figure out the meaning of events if for instance you could still have the "stopped" event sitting unfetched in the event queue, but the process state were shown as stopped.  To ensure this, we designate one of the listeners (the one passed into the Process constructor) as the "primary process event listener".  That one gets sent the event first, and we're supposed to only change the process state when the primary listener has finished handling the event.  Note, while a listener that hijacks the process event broadcaster gets designated as the primary event listener while it is hijacking the broadcaster, the converse is not true, the primary listener does not have to be hijacking the process event stream.  In fact, it's usually just the listener you passed in when you created the Process.

So why doesn't that all just work?

The problem is that when lldb goes to run the DoOnRemoval callback for the Process event (that's what calls your python breakpoint code, for instance), the code needs to see the process state as stopped - so that any commands will execute in the right context.  Currently, the only way to do that is to switch the public state to stopped.  But as you correctly analyzed, if you set the public state to stopped while we're in the middle of handling the stop, then any other code that's waiting on the state change will trigger, possibly before the breakpoint has decided whether to stop or not, which is incorrect.  To really honor the contract above, we need to make sure the stop event doesn't get delivered to the non-primary listeners till the stop is fully handled.

We already have code to deal with a slightly different aspect of this.  

The other reason that process events are a bit special is that we know that to implement a "user level operation" like source line stepping, the debugger often has to stop and start numerous times.  Those stops have no meaning at the user level, so we want to suppress them.  The way do that is to have a two stage handling of process events and their corresponding "state".  First the "private" side of lldb fetches the event (usually on the private state thread), and sets the "private state" to stopped without changing the "public state".  Then we run the thread plan logic is run to decide whether we are indeed going to stop to the user.  Then we rebroadcast the event to the primary process listener - and as a result set the "public state" to stopped.  But the code that runs the "Thread Plans" on the private state thread needs to see the state as stopped.

We implement this by making Process::GetState smart about who is using it.  If you look at Process::GetState, you will see it doesn't just return the public state.  That's because all the thread plan logic has to run on the Private State Thread before it is appropriate to set the public state to stopped.  Instead, when code running on the "Private State Thread" asks for the process state, it will get the private state.  But code running on any other thread will still see the public state. That's done by checking "CurrentThreadIsPrivateStateThread" in Process::GetState.

I think the correct solution to this problem is to add the thread that's going to run DoOnRemoval for the primary process event listener to the private state thread so that both of them will see the private state in Process::GetState.  Then we only flip the public state when DoOnRemoval is all the way done .  That way - like with the private event handling - the code doing the handling will see the right process state, so it will function correctly.  And then only when the callback has completed will the public state get switched (causing the event to be broadcast to the other clients waiting on the change in public state.)

That way the thread that is running the breakpoint command will run uninterrupted but in the correct context without any of the other threads waiting on the process knowing about the stop too early.

And that way the non-primary process event listeners will be purely advisory as they are supposed to be.

labath commented:
I think I understand most of what you're saying, and I think the proposed solution may solve the problem you have in mind -- but I don't think it can solve the problem *I* have in mind (and I'm not sure which of those is OP's problem :P). The race I *think* I see is in `Process::StopForDestroyOrDetach`.

We first check the process state (both of them), and if it's running, we install the hijack listener (and wait on it). However if the stopped event is sent (enqueued to the primary listener)  before the hijack listener is installed, the listener will never get anything. I don't see how fiddling with thread identities can help there. Even if it does, it will only help for one particular thread. OP is calling Kill from the event handling thread, but I don't think we require that. If Kill is called from some other thread, I think it could still run into the same bug (although the race window will most likely be smaller, since the event thread will be alive and pulling events from the queue).

jimingham commented:
We'd have to make the change and see to be sure, but what I would expect is that if any other thread tried to call Kill while whoever is the currently the primary process listener is handling the stop, it would immediately fail because the process is not stopped.  Right now, the Kill is sometimes allowed because the public state is "stopped" even though we haven't actually finished handling the stop event.  So I think what I was suggesting will fix the original problem.

The race Pavel is pointing out also seems real to me, but that should be fixed at the source, rather than after the fact. Process::HijackProcessEvents should check the previous "primary event listener" to see if it has any events and if so they should be transferred to the Hijack Listener.  We should also do the same thing in RestoreProcessEvents.  That way we'll never drop an event when shuffling primary listeners.

athierry-oct committed:
[lldb] Transfer events from previous listener when hijacking

This patch addresses a race condition that can occur when a script using
the Python API calls SBProcess::Kill() around the same time a breakpoint
is hit.

If a stop event is broadcast before the hijack listener is installed, it
may be delivered to the default listener and missed entirely by the
hijack listener. This causes Process::WaitForProcessToStop() to hang
until it times out, waiting for a public stop event that is never
received.

To fix this, we now transfer all pending events from the original
listener to the hijack listener before hijacking, using a new
Listener::MoveEvents method. We also do the reverse operation when
restoring the original listener to avoid dropping any event.

athierry-oct commented:
I've just pushed a new version that moves the pending events from the previous listener to the hijack listener when hijacking, and vice-versa when restoring the previous listener.

In the patch:

- Transferring the events between listeners is done atomically with adding/removing the hijack listener (by holding `m_listeners_mutex`), It should not be possible for an event to be received in the previous listener after the transfer of events but before the switch of listeners, because `m_listeners_mutex` is also locked by `PrivateBroadcastEvent()`
- In `HijackBroadcaster()` and `RestoreBroadcaster()`, we're holding both `m_listeners_mutex` and the listeners' `m_events_mutex`. So in theory, a deadlock is possible if some other thread locks `m_events_mutex` and then `m_listeners_mutex`. I've not seen any place in the code where that happens, but please let me know if I missed something.

jimingham reviewed (approved):
I don't think we ever have a use case where the Hijack listener would only want to see "new" events after the Hijack point.  And that's certainly not what we want for process events where the full sequence is important.  So the behavior you're encoding here is correct for the ways we use Listeners.  

I had to think a second about whether we wanted the notification behavior of AddEvent (it both adds the event to the queue and wakes up any Listeners).  But you have to do that.  For instance, the original listener is very likely sitting in WaitForEvent - it has no way of knowing it's been hijacked - and so the events you transferred back have to wake it up from the Wait or we'll stall.

jimingham commented:
I don't think the locking should be a real issue.  You should only hold the m_events mutex for a listener when adding or removing an event from the Listener's event queue.  But it doesn't make sense to actually do work when handing off the event.  

For instance, the primary place where events get pulled is Listener::FindNextEventInternal.  That function does trigger arbitrary work to be done after the event is fetched - in the DoOnRemoval call.  Since that's just a callout into a plugin provided callback we can't guarantee it won't do something that ends up requiring the listeners mutex.

But we are careful to do this like:

    if (remove) {
      m_events.erase(pos);
      // Unlock the event queue here.  We've removed this event and are about
      // to return it so it should be okay to get the next event off the queue
      // here - and it might be useful to do that in the "DoOnRemoval".
      lock.unlock();
      event_sp->DoOnRemoval();
    }

where that `lock` was the lock_guard for the m_events_mutex, so that the actual work is done on fetching the event is done after we drop the m_events_mutex.

I can't think of a good way to enforce this, but I'm also not sure that's necessary.  However, it might be a good idea to leave a comment by the definition of m_events_mutex saying that this should only be held when adding or fetching events from the queue, and not over any function that does arbitrary work on the event's behalf.

labath reviewed (commented):
Looks good, but you should add a unit test for the moving functionality. You know, things like: queue, some events, hijack them, make sure they're received at the new listener, and vice versa.

athierry-oct commented:
@jimingham Thanks for the explanation!

@labath I've added some unit tests, let me know if you need something else

labath reviewed (approved):
That's pretty much what I had in mind. Thanks.

athierry-oct commented:
@JDevlieghere done!

jimingham commented:
It doesn't look like there are any outstanding emendations.  Pushing.

jimingham commented:
The revert was:

https://github.com/llvm/llvm-project/pull/148915

Good hunting!

athierry-oct commented:
Hi, sorry for the delay. I investigated the `TestCallThatRestarts.py` failure, and I think I’ve figured out what’s going on.

Toward the end of the test, we run:

```python
value = frame.EvaluateExpression("call_me (%d)" % (num_sigchld), options)
```

Then we call:

```python
error = process.Continue()
```

This triggers `Process::ResumeSynchronous()`, which:
- Hijacks the process events
- Calls `PrivateResume()`
- Waits for a stop event via `WaitForProcessToStop()`

The issue is that a public stop event is sent at the end of `EvaluateExpression()`, and that event is still sitting in the primary listener’s queue when `Process::ResumeSynchronous()` hijacks the events. With my changes, that old stop event gets moved to the hijacker’s queue. So `ResumeSynchronous()` ends up grabbing it (even though it happened *before* the resume) and returns too early.

It looks like moving pending events during hijacking might not always be the right thing to do. In the case of `ResumeSynchronous()`, I think we want to make sure the stop event we wait for happens *after* hijacking and resuming.

One idea: we could add a `bool move_pending_events` flag to `HijackProcessEvents()` and `RestoreProcessEvents()`. It would default to `false`, and we could set it to `true` in `StopForDestroyOrDetach()`. This way, only the behavior of `StopForDestroyOrDetach()` is modified for now.

I gave this a quick try, and now the `check-lldb-api` suite passes on my machine.

Does that approach sound reasonable to you?

jimingham commented:
Seems to me the problem there was how EvaluateExpression knew that the expression was successfully or unsuccessfully completed without pulling that public Stop event off the event queue?  It shouldn't be possible to strand events like that.

athierry-oct commented:
Sorry for the long delay, I'm back from vacation and tackling this issue in parallel with other tasks.

I've started investigating the other failing tests (such as the DAP tests). For all failing tests, it seems essentially the root cause is the same: moving events between primary and hijacked listeners exposes places where events were left unprocessed in the primary listener or hijack listener queue. 

Focusing first on `EvaluateExpression()` and `TestCallThatRestarts`, the test failure is due to a stop event being rebroadcast to the primary listener when the process stops during expression evaluation, and then being unexpectedly consumed by `process.Continue()` (see previous messages)
To solve this, at first I thought maybe this stop event should be manually consumed in the test after the call to `frame.EvaluateExpression()` and before the call to `process.Continue()`
But now I'm realizing, since the tests are running in sync mode without an event loop, it seems this event should not have to be handled explicitly by the user (ie. the python test script in our case). 

I also came across the code of `SBDebugger::HandleCommand()`, and noticed that in sync mode it consumes events after running the command:

```cpp
void SBDebugger::HandleCommand(const char *command) {
  LLDB_INSTRUMENT_VA(this, command);

  ... // Handle the command

   // In sync mode, consume events produced by the command

    if (!m_opaque_sp->GetAsyncExecution()) {
      SBProcess process(GetCommandInterpreter().GetProcess());
      ProcessSP process_sp(process.GetSP());
      if (process_sp) {
        EventSP event_sp;
        ListenerSP lldb_listener_sp = m_opaque_sp->GetListener();
        while (lldb_listener_sp->GetEventForBroadcaster(
            process_sp.get(), event_sp, std::chrono::seconds(0))) {
          SBEvent event(event_sp);
          HandleProcessEvent(process, event, GetOutputFile(), GetErrorFile());
        }
      }
    }
  }
}
```

Would it make sense to do something similar for `SBFrame::EvaluateExpression()` ? This way, events would still be handled transparently when evaluating an expression in sync mode.

labath commented:
I think it'd be better to not do that. HandleCommand does that because it has no idea what events are going to be produced by the command it runs. In case of EvaluateExpression, it feels like there ought to be a way which events are going to be produced and consumed, and I think it'd be better to handle this where the events are being processed (when we're listening on the hijack listener) -- instead of letting the hijack listener forward events to who-knows-where, and then trying to clean up after it.

jimingham commented:
RunThreadPlan has to handle two different use cases.  In then normal case where -i is 1 and -u is 0, the expression evaluation always pretends that nothing happened in the target.  The stop-reason is set back to whatever it was before expression evaluation and no events are generated.  After all, expressions can be IR interpreted so their success doesn't necessarily depend on running the process.

But when say -i is 0 and you hit a breakpoint while running the expression, then we need to promote that breakpoint-caused stop event to the user so the process state will change to "stopped at a breakpoint in the middle of this expression".  If you are in sync mode, then that stop event should get consumed on the way out of EvaluateExpression, but in async mode that event should be propagated so that the process event listener can fetch the event and so the public state gets correctly reset.
